cmd/entrypoint/webhooks.go

```diff
@@ -0,0 +1,275 @@
+)
+
+const WEBHOOK_PORT int = 8043
+// TODO: automatic cert regeneration
+const CERT_VALID_DAYS int = 365
+const CA_SECRET_NAME string = "emissary-ingress-webhook-ca"
+
+func handleWebhooks(ctx context.Context) error {
+	scheme := runtime.NewScheme()
+
+	utilruntime.Must(clientgoscheme.AddToScheme(scheme))
+	utilruntime.Must(v2.AddToScheme(scheme))
+	utilruntime.Must(v3alpha1.AddToScheme(scheme))
+
+	// Create the webhook server
+	webhook := &conversion.Webhook{}
+	webhook.InjectScheme(scheme)
+	mux := http.NewServeMux()
+	mux.HandleFunc("/crdconvert", webhook.ServeHTTP)
+
+	// need a k8s client
+	kubeinfo := k8s.NewKubeInfo("", "", "")
+	restConfig, err := kubeinfo.GetRestConfig()
+	if err != nil {
+		return err
+	}
+	coreClient, err := k8sClientCoreV1.NewForConfig(restConfig)
+	if err != nil {
+		return err
+	}
+
+	// get CA secret
+	secrets := coreClient.Secrets(GetAmbassadorNamespace())
+	caSecret, err := secrets.Get(ctx, CA_SECRET_NAME, k8sTypesMetaV1.GetOptions{})
+	var caPEM *b.Buffer
+	var caTemplate *x509.Certificate
+	var caPrivKey *rsa.PrivateKey
+	if err != nil {
+		// Error here most likely means the secret doesnt exist yet
+		// (then we make one). Otherwise if its an actual error we exit here
+		if !k8sErrors.IsNotFound(err) {
+			return err
+		}
+
+		// Generate CA Certificate and key...
+		notBefore := time.Now()
+		notAfter := notBefore.Add(time.Duration(CERT_VALID_DAYS*24)*time.Hour)
+		serialNumberLimit := new(big.Int).Lsh(big.NewInt(1), 128)
+		serialNumber, err := rand.Int(rand.Reader, serialNumberLimit)
+		if err != nil {
+			return err
+		}
+		caTemplate = &x509.Certificate{
+			SerialNumber: serialNumber,
+			Subject: pkix.Name{
+				Organization:  []string{"Ambassador Labs"},
+			},
+			NotBefore:             notBefore,
+			NotAfter:              notAfter,
+			IsCA:                  true,
+			ExtKeyUsage:           []x509.ExtKeyUsage{x509.ExtKeyUsageClientAuth, x509.ExtKeyUsageServerAuth},
+			KeyUsage:              x509.KeyUsageDigitalSignature | x509.KeyUsageCertSign,
+			BasicConstraintsValid: true,
+		}
+		caPrivKey, err = rsa.GenerateKey(rand.Reader, 4096)
+		if err != nil {
+			return err
+		}
+		caBytes, err := x509.CreateCertificate(rand.Reader, caTemplate, caTemplate, &caPrivKey.PublicKey, caPrivKey)
+		if err != nil {
+			return err
+		}
+
+		// PEM encode certificate and key
+		caPEM = new(b.Buffer)
+		pem.Encode(caPEM, &pem.Block{
+			Type:  "CERTIFICATE",
+			Bytes: caBytes,
+		})
+		caPrivKeyPEM := new(b.Buffer)
+		pem.Encode(caPrivKeyPEM, &pem.Block{
+			Type:  "RSA PRIVATE KEY",
+			Bytes: x509.MarshalPKCS1PrivateKey(caPrivKey),
+		})
+
+		// Create and write the secret
+		_, err = secrets.Create(ctx, &k8sTypesCoreV1.Secret{
+			ObjectMeta: k8sTypesMetaV1.ObjectMeta{
+				Name: CA_SECRET_NAME,
+				Namespace: GetAmbassadorNamespace(),
+			},
+			Type: k8sTypesCoreV1.SecretTypeTLS,
+			Data: map[string][]byte{
+				"tls.key": caPrivKeyPEM.Bytes(),
+				"tls.crt": caPEM.Bytes(),
```

Comment:
Nit: Style: Prefer to use the `corev1.TLSCertKey` and `corev1.TLSPrivateKeyKey` constants instead of string literals.